src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+        oldNode.value = value;
+        return oldNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> nodeToRemove = first;
+        while (nodeToRemove != null) {
+            if (nodeToRemove.value == object
+                    || (nodeToRemove.value != null && nodeToRemove.value.equals(object))) {
+                unlink(nodeToRemove);
+                return true;
+            }
+            nodeToRemove = nodeToRemove.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void validateIndexToAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void validateIndexToGet(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(index + " index is wrong for size : " + size);
+        }
+    }
+
+    private void addLast(T value) {
+        Node<T> tempLast = last;
+        Node<T> newNode = new Node<>(tempLast, value, null);
+        last = newNode;
+        if (tempLast == null) {
+            first = newNode;
+        } else {
+            tempLast.next = last;
+        }
+        size++;
+    }
+
+    private void addFirst(T value, Node<T> node) {
+        Node<T> nodeBefore = node.prev;
+        Node<T> newNode = new Node<>(nodeBefore, value, node);
+        if (nodeBefore == null) {
+            first = newNode;
+        } else {
+            nodeBefore.next = newNode;
+            node.prev = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        validateIndexToGet(index);
+        Node<T> temp;
+        if (index < size / 2) {
+            temp = first;
+            for (int i = 0; i < index; i++) {
+                temp = temp.next;
+            }
+        } else {
+            temp = last;
+            for (int i = size - 1; i > index; i--) {
+                temp = temp.prev;
+            }
+        }
+        return temp;
+    }
+
+    private void unlink(Node<T> nodeToRemove) {
+        if (nodeToRemove.prev == null) {
+            first = nodeToRemove.next;
+        } else {
+            nodeToRemove.prev.next = nodeToRemove.next;
+        }
+        if (nodeToRemove.next == null) {
+            last = nodeToRemove.prev;
+        } else {
+            nodeToRemove.next.prev = nodeToRemove.prev;
+        }
+        size--;
+    }
+
+    private static class Node<T> {
+        private Node<T> next;
+        private Node<T> prev;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
```

Comment:
```suggestion
        Node(Node<T> prev, T value, Node<T> next) {
```
it won't be accessible anyway, so better leave default or put private for the clarity

Reply:
done